src/storage/on_disk.rs

```diff
@@ -0,0 +1,242 @@
+            .open_table(Some("block_hashes"))
+            .and_then(|table| tx.get(&table, B64::from(*number).as_ref()))
+            .map(|bytes: Option<[u8; 32]>| bytes.map(B256::from))?
+            .unwrap_or_else(|| keccak256(number.to_string().as_bytes()));
+
+        self.cache_block_hashes
+            .lock()
+            .unwrap()
+            .insert(*number, block_hash);
+        Ok(block_hash)
+    }
+}
+
+const MB: isize = 1048576;
+
+#[allow(clippy::identity_op)]
+const DEFAULT_DB_OPTIONS: DatabaseOptions = DatabaseOptions {
+    max_tables: Some(16),
+    mode: Mode::ReadWrite(ReadWriteOptions {
+        // https://erthink.github.io/libmdbx/group__c__settings.html#ga79065e4f3c5fb2ad37a52b59224d583e
+        // https://github.com/erthink/libmdbx/issues/136#issuecomment-727490550
+        sync_mode: SyncMode::Durable,
+        min_size: Some(1 * MB), // The lower bound allows you to prevent database shrinking below certain reasonable size to avoid unnecessary resizing costs.
+        max_size: Some(1024 * MB), // The upper bound allows you to prevent database growth above certain reasonable size.
+        growth_step: Some(1 * MB), // The growth step must be greater than zero to allow the database to grow, but also reasonable not too small, since increasing the size by little steps will result a large overhead.
+        shrink_threshold: Some(4 * MB), // The shrink threshold must be greater than zero to allow the database to shrink but also reasonable not too small (to avoid extra overhead) and not less than growth step to avoid up-and-down flouncing.
+    }),
+    permissions: None,
+    max_readers: None,
+    rp_augment_limit: None,
+    loose_limit: None,
+    dp_reserve_limit: None,
+    txn_dp_limit: None,
+    spill_max_denominator: None,
+    spill_min_denominator: None,
+    page_size: None,
+    no_sub_dir: false,
+    exclusive: false,
+    accede: false,
+    no_rdahead: false,
+    no_meminit: false,
+    coalesce: false,
+    liforeclaim: false,
+};
+
+fn open_db(dir: impl AsRef<Path>) -> Result<Database<NoWriteMap>, libmdbx::Error> {
+    Database::<NoWriteMap>::open_with_options(dir.as_ref(), DEFAULT_DB_OPTIONS)
+}
+
+fn write_table_to<E: DatabaseKind, K: AsRef<[u8]>, V: AsRef<[u8]>>(
+    db: &Database<E>,
+    table_name: &str,
+    entries: impl Iterator<Item = (K, V)>,
+) -> Result<(), libmdbx::Error> {
+    let tx = db.begin_rw_txn()?;
+    let table = tx.create_table(Some(table_name), TableFlags::default())?;
+    for (k, v) in entries {
+        tx.put(&table, k, v, WriteFlags::UPSERT)?;
+    }
+    tx.commit()?;
+    Ok(())
+}
+
+/// Create a temp dir containing MDBX
+pub fn create_db_dir<'a>(
+    block_number: &str,
+    bytecodes: impl Iterator<Item = (&'a B256, &'a EvmCode)>,
+    pre_state: impl Iterator<Item = (&'a Address, &'a EvmAccount)>,
+    block_hashes: impl Iterator<Item = (&'a u64, &'a B256)>,
+) -> Result<PathBuf, libmdbx::Error> {
+    let mut dir = std::env::temp_dir();
```

Comment:
Why are these defined in library source instead of `tests/common`? This usage seems too niche to be useful for users. And we can always extract it back from tests when there is a clear proposed use case.

Reply:
It's pretty big so we can just put it in `tests/common/mdbx.rs`.